xds/src/main/java/io/grpc/xds/internal/Matchers.java

```diff
@@ -0,0 +1,332 @@
+    }
+
+    /** The request header value should be within the range. */
+    public static HeaderMatcher forRange(String name, Range range, boolean inverted) {
+      checkNotNull(name, "name");
+      checkNotNull(range, "range");
+      return HeaderMatcher.create(name, null, null, range, null, null, null, inverted);
+    }
+
+    /** The request header value should exist. */
+    public static HeaderMatcher forPresent(String name, boolean present, boolean inverted) {
+      checkNotNull(name, "name");
+      return HeaderMatcher.create(name, null, null, null, present, null, null, inverted);
+    }
+
+    /** The request header value should have this prefix. */
+    public static HeaderMatcher forPrefix(String name, String prefix, boolean inverted) {
+      checkNotNull(name, "name");
+      checkNotNull(prefix, "prefix");
+      return HeaderMatcher.create(name, null, null, null, null, prefix, null, inverted);
+    }
+
+    /** The request header value should have this suffix. */
+    public static HeaderMatcher forSuffix(String name, String suffix, boolean inverted) {
+      checkNotNull(name, "name");
+      checkNotNull(suffix, "suffix");
+      return HeaderMatcher.create(name, null, null, null, null, null, suffix, inverted);
+    }
+
+    private static HeaderMatcher create(String name, @Nullable String exactValue,
+        @Nullable Pattern safeRegEx, @Nullable Range range,
+        @Nullable Boolean present, @Nullable String prefix,
+        @Nullable String suffix, boolean inverted) {
+      checkNotNull(name, "name");
+      return new AutoValue_Matchers_HeaderMatcher(name, exactValue, safeRegEx, range, present,
+          prefix, suffix, inverted);
+    }
+
+    /** Represents an integer range. */
+    @AutoValue
+    public abstract static class Range {
+      public abstract long start();
+
+      public abstract long end();
+
+      public static Range create(long start, long end) {
+        return new AutoValue_Matchers_HeaderMatcher_Range(start, end);
+      }
+    }
+
+    @Nullable
+    private static String getHeaderValue(Metadata headers, String headerName) {
+      if (headerName.endsWith(Metadata.BINARY_HEADER_SUFFIX)) {
+        return null;
+      }
+      if (headerName.equals("content-type")) {
+        return "application/grpc";
+      }
+      Metadata.Key<String> key;
+      try {
+        key = Metadata.Key.of(headerName, Metadata.ASCII_STRING_MARSHALLER);
+      } catch (IllegalArgumentException e) {
+        return null;
+      }
+      Iterable<String> values = headers.getAll(key);
+      return values == null ? null : Joiner.on(",").join(values);
+    }
+
+    /** Returns the matching result. */
+    public boolean matches(Metadata headers) {
+      checkNotNull(headers);
+      String value = getHeaderValue(headers, name());
+      if (present() != null) {
+        return (value == null) == present().equals(inverted());
```

Comment:
Avoid Boolean boxing? `present().booleanValue() == inverted()`

Reply:
Sorry, I didn't realize this was preexisting. We can leave it as it was. Note: I think the `value == null` case is broken because it doesn't check inverted(), but let's do any fixups there as a separate PR.

Reply:
Why it is broken? Except I forgot to put a `@Nullable` for the method argument...

The spec of this method is (would be):
Given the value of a header with name equal to this HeaderMatcher's name, returns `true` if this matcher matches the header. The given header value would be `null` if such a header doesn't exist/present.

So this line of code handles the cases of presence matching:



|header exists| presence requirement | inverted bit | result | comment |
|---|---|---|---|---|
| false | false | false | match | matcher requires the header _must be absent_ |
| false | false | true | not match | matcher requires the header _must be absent_, but inverted bit is set, so it requires the header _must be present_ |
| false | true | false | not match | matcher requires the header _must be present_ |
| false | true | true | match | matcher requires the header _must be present_, but inverted bit is set, so it requires the header _must be absent_ |
| true | false | false | not match | matcher requires the header _must be absent_ |
| true | false | true | match | matcher requires the header _must be absent_, but inverted bit is set, so it requires the header _must be present_ |
| true | true | false | match | matcher requires the header _must be present_ |
| true | true | true | not match | matcher requires the header _must be present_, but inverted bit is set, so it requires the header _must be absent_ |

So based on the truth table, this result is XNOR of `value != null`, `present() == true` and `inverted() == true`.

Reply:
Oh, I do have the `@Nullable` in the [original code](https://github.com/grpc/grpc-java/blob/fa4b980e07400a66cfaf3693785f873031b20482/xds/src/main/java/io/grpc/xds/XdsNameResolver.java#L580). So this has been refactored and the spec of this method might have changed. Let me take a deeper look.

Reply:
Looks the spec is relatively the same for this method and that in [XdsNameResolver](https://github.com/grpc/grpc-java/blob/fa4b980e07400a66cfaf3693785f873031b20482/xds/src/main/java/io/grpc/xds/XdsNameResolver.java#L580), in which `value` should be `Nullable`.

Btw, now that you moved the implementation of `matchHeader(...)` here, you need to delete the copy in `XdsNameResolver` and call `HeaderMatcher.matches(...)` instead.

Reply:
For this HeaderMatcher.matches() method it was a copy of the one in XdsNameResolver. The key/value lookup is different but the match logic looks the same, so yes it might be easily refactored to one place, but as we voted before it might also be acceptable for different users to implement their own matches() logic, regardless of minor code duplicates. Also I think it is fair to let this PR focus on what it immediately needs instead of refactoring ramifications here and there.

Reply:
Okay, what I meant was you could just simply replace the implementation of `XdsNameResolver.matchHeader(headerMatcher, value)` with `headerMatcher.matches(value)` if you do not want to do structural changes in this PR. But fine, I am ok with duplicated code in independent scopes.

The argument of this method should be annotated with `@Nullable`.